Provider/src/main/java/com/spotify/confidence/client/serializers/Serializers.kt

```diff
@@ -25,79 +23,103 @@ import kotlinx.serialization.json.Json
 import kotlinx.serialization.json.JsonDecoder
 import kotlinx.serialization.json.JsonElement
 import kotlinx.serialization.json.JsonNull
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.boolean
+import kotlinx.serialization.json.booleanOrNull
+import kotlinx.serialization.json.double
+import kotlinx.serialization.json.doubleOrNull
+import kotlinx.serialization.json.int
+import kotlinx.serialization.json.intOrNull
 import kotlinx.serialization.json.jsonArray
 import kotlinx.serialization.json.jsonObject
 import kotlinx.serialization.json.jsonPrimitive
-import java.util.UUID
+import kotlinx.serialization.serializer
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.TimeZone
 
 /***
  * the struct serializer needed for sending the resolve request
  */
 
-internal object StructureSerializer : KSerializer<Structure> {
+internal object StructSerializer : KSerializer<ConfidenceValue.Struct> {
     override val descriptor: SerialDescriptor =
         MapSerializer(String.serializer(), String.serializer()).descriptor
 
-    override fun deserialize(decoder: Decoder): Structure {
+    override fun deserialize(decoder: Decoder): ConfidenceValue.Struct {
         error("no deserializer is needed")
     }
 
-    override fun serialize(encoder: Encoder, value: Structure) {
+    override fun serialize(encoder: Encoder, value: ConfidenceValue.Struct) {
         encoder.encodeStructure(descriptor) {
-            for ((key, mapValue) in value.asMap()) {
+            for ((key, mapValue) in value.map) {
                 encodeStringElement(descriptor, 0, key)
-                encodeSerializableElement(descriptor, 1, StructureValueSerializer, mapValue)
+                encodeSerializableElement(descriptor, 1, ConfidenceValueSerializer, mapValue)
             }
         }
     }
 }
 
-private object StructureValueSerializer : KSerializer<Value> {
+internal object ConfidenceValueSerializer : KSerializer<ConfidenceValue> {
     override val descriptor: SerialDescriptor
         get() = ValueSerializer.descriptor
 
-    override fun deserialize(decoder: Decoder): Value {
-        error("we don't need to deserialize here")
+    override fun deserialize(decoder: Decoder): ConfidenceValue {
+        require(decoder is JsonDecoder)
+        when (val element = decoder.decodeJsonElement()) {
+            is JsonPrimitive -> {
+                val jsonPrimitive = element.jsonPrimitive
+                return when {
+                    jsonPrimitive.isString -> ConfidenceValue.String(jsonPrimitive.content)
+                    jsonPrimitive.booleanOrNull != null -> ConfidenceValue.Boolean(jsonPrimitive.boolean)
+                    jsonPrimitive.intOrNull != null -> ConfidenceValue.Integer(jsonPrimitive.int)
+                    jsonPrimitive.doubleOrNull != null -> ConfidenceValue.Double(jsonPrimitive.double)
+                    else -> ConfidenceValue.Null
+                }
```

Comment:
as discussed, these are pretty uncertain, let's defer until we've aligned on how to handle.

Reply:
updated to use the `Polymorphic` serialization like we do for `Value` in the sdk.